src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs

```diff
@@ -177,11 +175,11 @@ public void Dispose()
 
         private (uint RefEscapeScope, uint ValEscapeScope) GetLocalScopes(LocalSymbol local)
         {
-            Debug.Assert(_localEscapeScopes?.ContainsKey(local) == true);
+            Debug.Assert(_localEscapeScopes?.ContainsKey(local) == true || _symbol != local.ContainingSymbol);
 
             return _localEscapeScopes?.TryGetValue(local, out var scopes) == true
                 ? scopes
-                : (CallingMethodScope, CallingMethodScope);
+                : (CurrentMethodScope, CallingMethodScope);
```

Comment:
It looks like this has the effect of always saying that a local in a method/function has these ref/value lifetimes? Is that safe?

I think this is OK for the value lifetime because you can't capture a ref struct in a lambda. But I'm not so sure for the ref lifetime. Trying to think of a test.

Reply:
I think I am comfortable with the captured local ref having `CurrentMethod` lifetime.

It feels like in reality, the captured local has some "external" lifetime relative to the current function. It is always still live after the local function call returns. It does compile down to a ref parameter, after all, when the function is not converted to delegate, etc.

But, it is safe for us to, in the context of a capture, implicitly convert that captured local's lifetime to CurrentMethod of the current function, since converting from a wider to narrower top-level lifetime is generally safe.

It means the captured local might not be convertible to some things that it maybe ought to be, like, a ref parameter or return, but this doesn't seem like a coding pattern people really need. Imagine something like the following, which I think is safe, but that can't be done as written, either before or after this PR.

```cs
bool flag = false;
int x = 0;
int y = 0;
ref int z = fn();
ref int fn()
    => ref flag ? ref x : ref y;
```

I think also if we wanted the ability to capture ref struct locals, we would have to reckon with this problem a little more, but I am also not seeing that as something people are urgently asking to do, so am not worried about it.

Reply:
Right. From implementation perspective, the captured locals could have the CallingMethod ref lifetime, and everything should work since the equivalent lowered code would be ref safe.

But from the language specification perspective, I'm not sure we can make that assumption. Perhaps there are alternative lowering strategies where the ref safety would be violated.

If we wanted to change this, I think we would need to change the ref safety speclet to specify lifetimes of captured locals (or propose a new feature for that).